source/s3_checksum_stream.c

```diff
@@ -14,10 +14,29 @@ struct aws_checksum_stream {
     struct aws_input_stream *old_stream;
     struct aws_s3_checksum *checksum;
     struct aws_byte_buf checksum_result;
-    /* base64 encoded checksum of the stream, updated on destruction of stream */
+    /* base64 encoded checksum of the stream, updated at end of stream */
     struct aws_byte_buf *encoded_checksum_output;
+    bool checksum_finalized;
 };
 
+int s_aws_finalize_checksum(struct aws_checksum_stream *impl) {
+    if (impl->checksum_finalized) {
+        return AWS_OP_SUCCESS;
+    }
+
+    int result = aws_checksum_finalize(impl->checksum, &impl->checksum_result, 0);
+    if (result != AWS_OP_SUCCESS) {
+        aws_byte_buf_reset(&impl->checksum_result, true);
+        impl->checksum_finalized = true;
+        return aws_raise_error(AWS_ERROR_S3_CHECKSUM_CALCULATION_FAILED);
```

Comment:
In general, I would expect all the function call, `aws_checksum_finalize`, will raise the proper error if the result is not AWS_OP_SUCCESS, so that the caller don't need to raise another error.

I looked into `aws_checksum_finalize`, it seems to raise lot of different errors and may not be very informative. Yeah, we should override the error code here, but I'd also suggest to log the previous error code for better error tracking.